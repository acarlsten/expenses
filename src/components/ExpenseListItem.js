import React from 'react'
import { Link } from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'
import 'numeral/locales/no'
numeral.locale('no') // same rules as SE

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>Summa: {numeral(amount / 100).format('0,0[.]00 $')}</p>
    <p>Skapad: {moment(createdAt).format('LL')}</p>
  </div>
)

export default ExpenseListItem
