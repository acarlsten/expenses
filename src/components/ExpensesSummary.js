import React from 'react'
import { connect } from 'react-redux'
import selectExpenses from '../selectors/expenses'
import getExpensesTotal from '../selectors/expenses-total'
import numeral from 'numeral'
import 'numeral/locales/no'
numeral.locale('no') // same rules as SE

export const ExpensesSummary = props => (
  <div>
    {props.expensesCount === 0 ? (
      <p>Nothing to see here!</p>
    ) : (
      <p>
        Viewing {props.expensesCount} {props.expensesCount === 1 ? 'expense' : 'expenses'} totalling{' '}
        {numeral(props.expensesTotal / 100).format('0,0[.]00 $')}
      </p>
    )}
  </div>
)

const mapStateToProps = state => {
  return {
    expensesTotal: getExpensesTotal(selectExpenses(state.expenses, state.filters)),
    expensesCount: selectExpenses(state.expenses, state.filters).length
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
