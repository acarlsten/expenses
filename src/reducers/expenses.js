const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense]
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => action.id !== id)
    case 'EDIT_EXPENSE':
      return state.map(expense => {
        // trying to change id causes silent failure, for now
        if (expense.id === action.id && !action.updates.id) {
          return {
            ...expense,
            ...action.updates
          }
        } else {
          return expense
        }
      })
    case 'SET_EXPENSES':
      return action.expenses
    default:
      return state
  }
}
