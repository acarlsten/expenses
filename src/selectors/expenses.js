import moment from 'moment'
//get visible expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt)
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

      return startDateMatch && endDateMatch && textMatch
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1
      } else if (sortBy === 'amount') {
        if (a.amount > b.amount) {
          return -1
        } else if (a.amount < b.amount) {
          return 1
        }
        // if the amounts are equal check and sort by latest date
        if (a.createdAt < b.createdAt) {
          return 1
        } else if (a.createdAt > b.createdAt) {
          return -1
        } else {
          return 0
        }
      }
    })
}
