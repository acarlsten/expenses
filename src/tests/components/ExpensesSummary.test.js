import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'
import expenses from '../fixtures/expenses'
import getExpensesTotal from '../../selectors/expenses-total'

test('should render ExpenseSummary with no expenses', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={0} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseSummary with one expense', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={expenses[0].amount} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseSummary with several expenses', () => {
  const wrapper = shallow(
    <ExpensesSummary expensesCount={expenses.length} expensesTotal={getExpensesTotal(expenses)} />
  )
  expect(wrapper).toMatchSnapshot()
})
