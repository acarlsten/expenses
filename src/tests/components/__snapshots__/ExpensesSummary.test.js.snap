// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with no expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        0
      </span>
       expenses. There are 
       expenses in total.
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary with one expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <div>
      <h1
        className="page-header__title"
      >
        Viewing 
        <span>
          1
        </span>
         
        expense
         totalling
         
        <span>
          1,95 kr
          .
        </span>
      </h1>
      <p
        className="page-header__subtitle"
      >
        There are 
        <span />
         expenses in total.
      </p>
    </div>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSummary with several expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <div>
      <h1
        className="page-header__title"
      >
        Viewing 
        <span>
          3
        </span>
         
        expenses
         totalling
         
        <span>
          1 141,95 kr
          .
        </span>
      </h1>
      <p
        className="page-header__subtitle"
      >
        There are 
        <span />
         expenses in total.
      </p>
    </div>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
